src/file_builder.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+use std::io::{Read, Seek, Write};
+use std::path::Path;
+
+/// A `FileBuilder` generates file readers or writers that are composed upon
+/// existing ones.
+pub trait FileBuilder: Send + Sync {
+    /// Types of outcome file reader/writer. They must not alter the length of
+    /// bytes stream, nor buffer bytes between operations.
+    type Reader<R: Seek + Read>: Seek + Read;
```

Comment:
`type Reader: Seek + Read` should be enough? Then all across the file builder should use `Reader`.

Reply:
It is enough, but then we can't guarantee the implementor will actually use `R`/`W` from typing perspective. Though I'm not sure this is the idiomatic way to contracting such requirement.

Reply:
Also, I think it would be hard for implementor to wrap a generic `R`/`W` into a non-generic `Reader` type.